src/data/matrix/block.lean

```diff
@@ -213,16 +213,49 @@ by { ext i, cases i; simp [vec_mul, dot_product] }
 
 variables [decidable_eq l] [decidable_eq m]
 
-@[simp] lemma from_blocks_diagonal [has_zero α] (d₁ : l → α) (d₂ : m → α) :
+section has_zero
+variables [has_zero α]
+
+lemma to_block_diagonal_eq (d : m → α) (p : m → Prop) :
+  matrix.to_block (diagonal d) p p = diagonal (λ i : subtype p, d ↑i) :=
+begin
+  ext i j,
+  by_cases i = j,
+  { simp [h] },
+  { simp [has_one.one, h, λ h', h $ subtype.ext h'], }
+end
+
+lemma to_block_diagonal_ne (d : m → α) (p : m → Prop) :
+  matrix.to_block (diagonal d) (λ i, ¬ p i) p = 0 :=
```

Comment:
This lemma has a bit of a weird name, as there's no ≠ anywhere in the statement.

Is this true more generally as:
```suggestion
  disjoint p q → matrix.to_block (diagonal d) p q = 0 :=
```

Reply:
Yes, that's a great idea. I've generalized and renamed to `to_block_diagonal_self` and `to_block_diagonal_disjoint`.